src/array/utf8/mutable.rs

```diff
@@ -425,6 +425,29 @@ impl<O: Offset> MutableUtf8Array<O> {
         // soundness: T: AsRef<str>
         unsafe { Self::from_data_unchecked(Self::default_data_type(), offsets, values, None) }
     }
+
+    /// Write values directly into the underlying data.
+    /// # Safety
+    /// Caller must ensure that `len <= self.capacity()`
+    #[inline]
+    pub unsafe fn write_values<F>(&mut self, mut f: F)
+    where
+        F: FnMut(&mut [u8]) -> usize,
+    {
+        // ensure values has enough capacity and size to write
+        self.values.set_len(self.values.capacity());
+        let buffer = &mut self.values.as_mut_slice()[self.offsets.last().unwrap().to_usize()..];
```

Comment:
I think that this is unsound, even in `unsafe` code: a slice must always have initialized data on it. I propose a different implementation below that avoids introducing another API to the `MutableBuffer`. LMK what you think.

Reply:
Yes, I do agree with you. But now `MutableBuffer ` is the only way to construct a BinaryArray,  maybe we should expose `values`, `offsets` to outside.

We can have temp `values`, `offset` vectors in the cast kernel and then construct the `MutableBuffer` by these two vectors.

Refer to clickhouse's style: 
https://github.com/ClickHouse/ClickHouse/blob/515cc74530d11e1b2b18a63141b66a15b94748ba/src/Columns/ColumnString.h